### PR TITLE
Add WireGuard tests for iOS app

### PR DESCRIPTION
**Pull request by niklasberglund**
In this PR the following tests have been implemented:
* `testWireGuardOverTCPManually` - https://linear.app/mullvad/issue/IOS-431/test-wireguard-over-tcp-manually
* `testWireGuardOverTCPAutomatically` - https://linear.app/mullvad/issue/IOS-429/test-wireguard-over-tcp-works-automatically
* `testWireGuardPortSettings` - https://linear.app/mullvad/issue/IOS-434/test-wireguard-port-settings

To test it out you can run the three tests, they are all under `RelayTests`. Another configuration property `SHOULD_BE_REACHABLE_DOMAIN = mullvad.net` has been introduced so make sure to set it in your `UITests.xcconfig` file. You must be connected to the `app-team-ios-tests` wifi to be able to run tests interacting with the firewall running on the NUC.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955)

### CODE REVIEW

**Review by niklasberglund — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955)* status: 0 of 19 files reviewed, 1 unresolved discussion

___
*[`ios/MullvadVPNUITests/RelayTests.swift` line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nsx5S36-9sQnC1BQLnd:-Nsx5S36-9sQnC1BQLne:b-fg00hq) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/52dc88cd1c3a58735dcb7963254536e1e6d933a2/ios/MullvadVPNUITests/RelayTests.swift#L16)):*
> ```Swift
>         super.setUp()
> 
>         FirewallAPIClient().removeRules()
> ```

Removing rules in both `setUp` and `tearDown` is a compromise. Because the firewall API cannot be reached when connected to VPN, and it's difficult and extends the time tests need to run if trying to make sure to always disconnect before invoking `removeRules()` from `tearDown`. In a perfect world `tearDown` would always make sure to disconnect from relay if connected.

**Review by niklasberglund — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955)* status: 0 of 19 files reviewed, 1 unresolved discussion

___
*[`ios/MullvadVPNUITests/RelayTests.swift` line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nsx5S36-9sQnC1BQLnd:-Nsx8C8G-JFs8dRK4yFJ:b-l17hvc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/52dc88cd1c3a58735dcb7963254536e1e6d933a2/ios/MullvadVPNUITests/RelayTests.swift#L16)):*
<details><summary><i>Previously, niklasberglund (Niklas Berglund) wrote…</i></summary><blockquote>

Removing rules in both `setUp` and `tearDown` is a compromise. Because the firewall API cannot be reached when connected to VPN, and it's difficult and extends the time tests need to run if trying to make sure to always disconnect before invoking `removeRules()` from `tearDown`. In a perfect world `tearDown` would always make sure to disconnect from relay if connected.
</blockquote></details>

Also `testWireGuardOverTCPAutomatically` is the only test in this suite creating a firewall rule but `removeRules()` runs for all. So there could be logic for only running it when needed. I was thinking it doesn't hurt to run it even when it isn't needed and it doesn't add much time but would be great to hear your thoughts.

**Review by buggmagnet — CHANGES_REQUESTED** [review]
> Reviewed 18 of 19 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955)* status: all files reviewed, 8 unresolved discussions (waiting on @niklasberglund)

___
*[`ios/MullvadVPNUITests/RelayTests.swift` line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nsx5S36-9sQnC1BQLnd:-Nt0Ge4w4K823z34G3Mg:bd32pyq) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/52dc88cd1c3a58735dcb7963254536e1e6d933a2/ios/MullvadVPNUITests/RelayTests.swift#L16)):*
<details><summary><i>Previously, niklasberglund (Niklas Berglund) wrote…</i></summary><blockquote>

Also `testWireGuardOverTCPAutomatically` is the only test in this suite creating a firewall rule but `removeRules()` runs for all. So there could be logic for only running it when needed. I was thinking it doesn't hurt to run it even when it isn't needed and it doesn't add much time but would be great to hear your thoughts.
</blockquote></details>

I've been thinking about this since the discussion we had yesterday. 
If the firewall API is running on a local network, it should still be reachable when the VPN is turned on, since the `excludeLocalNetworks` on the VPN connection on iOS is on by default (i.e. local traffic won't be routed through the VPN)
We should double check this, and take a closer look why your original test attempts didn't work.
___
*[`ios/MullvadVPNUITests/RelayTests.swift` line 80 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0I9o45D1ncdtgWx6f:-Nt0I9o45D1ncdtgWx6g:bo50q92) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/RelayTests.swift#L80)):*
> ```Swift
> 
>     func testWireGuardOverTCPAutomatically() throws {
>         let wireGuardGot001RelayIPAddress = "185.213.154.66"
> ```

We can't really do this because the server IP might change in the future.
It's okay to leave it for now, but what we should probably be doing is either leveraging the JSON file that contains all the relays IPs  that we automatically download when we build the app, or do something similar to avoid having a static IP defined.
___
*[`ios/MullvadVPNUITests/Networking/FirewallAPIClient.swift` line 83 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0J-tk2sNX1mw3jKML:-Nt0J-tk2sNX1mw3jKMM:bduxrm4) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75c053739a33b0314b13951a9dce04cfd1fb52da/ios/MullvadVPNUITests/Networking/FirewallAPIClient.swift#L83)):*
> ```Swift
>         var request = URLRequest(url: removeRulesURL)
>         request.httpMethod = "DELETE"
>         request.setValue("application/json", forHTTPHeaderField: "Content-Type")
> ```

Why did we remove that line ? What was not working ?
___
*[`ios/MullvadVPNUITests/Networking/Networking.swift` line 89 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0JhI14ge9AXFukO4N:-Nt0JhI14ge9AXFukO4O:b-pr9lpa) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Networking/Networking.swift#L89)):*
> ```Swift
> 
>     /// Get configured domain to use for Internet connectivity checks
>     private static func getShouldBeReachableDomain() throws -> String {
> ```

nit
The name doesn't read super well, I think we can rename this `alwaysReachableDomain` and get the same intended idea.

With this addition, the checks become more readable too IMHO
```swift
XCTAssertTrue(try canConnectSocket(host: alwaysReachableDomain(), port: "80"))
```

What do you think ?
___
*[`ios/MullvadVPNUITests/Networking/Networking.swift` line 154 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0KfoM2juaRflg2N-I:-Nt0KfoM2juaRflg2N-J:bscq3wm) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Networking/Networking.swift#L154)):*
> ```Swift
>     }
> 
>     /// Verify that the device do not have Internet connectivity
> ```

nit 
`does not have`
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 31 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0LihU44p2w4FkHgBb:-Nt0LihU44p2w4FkHgBc:b99fdls) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L31)):*
> ```Swift
>         var capturedTexts: [String] = []
> 
>         while -startTime.timeIntervalSinceNow < timeout {
> ```

I understand what you're trying to do, but it looks really convoluted.
It's not performance critical code, so we should do `Date().timeIntervalSince(startTime)` instead for readability's sake
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 103 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0ZaMW-IkoURR7pRxq:-Nt0ZaMW-IkoURR7pRxr:b-z0xx1j) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L103)):*
> ```Swift
> 
>     /// Verify that the app attempts to connect over UDP before switching to TCP. For testing blocked UDP traffic.
>     @discardableResult func verifyConnectingOverTCPAfterUDPAttempts() -> Self {
> ```

nit
The logic in this function is a bit hard to follow, so the comments are very appreciated.
I'm wondering if we can make it easier to follow. I'll think about it.
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 106 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0ZRUf19meNKTKx66D:-Nt0ZRUf19meNKTKx66E:b-iru6v8) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L106)):*
> ```Swift
>         let connectionAttempts = waitForConnectionAttempts(3, timeout: 15)
> 
>         var i = 0
> ```

Can we rename this to something more meaningful ?

**Review by rablador — CHANGES_REQUESTED** [review]
> Reviewed 18 of 19 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955)* status: all files reviewed, 15 unresolved discussions (waiting on @buggmagnet and @niklasberglund)

___
*[`ios/MullvadVPN/Coordinators/ApplicationCoordinator.swift` line 770 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0LInl4fbtsg2xXwuV:-Nt0LInl4fbtsg2xXwuW:b-9yx2yr) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPN/Coordinators/ApplicationCoordinator.swift#L770)):*
> ```Swift
> 
>         let navigationController = CustomNavigationController()
>         navigationController.view.accessibilityIdentifier = .settingsContainerView
> ```

How come this isn't set on the view controller inside `SettingsCoordinator` instead?
___
*[`ios/MullvadVPN/View controllers/Settings/SettingsViewController.swift` line 58 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0MdVf8F2UUTccUWAI:-Nt0MdVf8F2UUTccUWAJ:b-d5klvf) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPN/View%20controllers/Settings/SettingsViewController.swift#L58)):*
> ```Swift
>         )
>         doneButton.accessibilityIdentifier = .settingsDoneButton
>         navigationItem.rightBarButtonItem = doneButton
> ```

I know we discussed the issues with navigation bar buttons before, but do we really need to go through these hoops to make it work?
___
*[`ios/MullvadVPNUITests/RelayTests.swift` line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nsx5S36-9sQnC1BQLnd:-Nt0NBuV9UeXwr3-osDe:b-izop0e) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/52dc88cd1c3a58735dcb7963254536e1e6d933a2/ios/MullvadVPNUITests/RelayTests.swift#L16)):*
<details><summary><i>Previously, niklasberglund (Niklas Berglund) wrote…</i></summary><blockquote>

Also `testWireGuardOverTCPAutomatically` is the only test in this suite creating a firewall rule but `removeRules()` runs for all. So there could be logic for only running it when needed. I was thinking it doesn't hurt to run it even when it isn't needed and it doesn't add much time but would be great to hear your thoughts.
</blockquote></details>

I'd say run `removeRules()` only when you needed, unless it's needed often enough to warrant a general inclusion. In general, I think we shouldn't proactively add stuff that we don't actually use.
___
*[`ios/MullvadVPNUITests/RelayTests.swift` line 59 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0Q2EO4YPTrUMQyY8x:-Nt0Q2EO4YPTrUMQyY8y:b-c73h6a) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/RelayTests.swift#L59)):*
> ```Swift
>         VPNSettingsPage(app)
>             .tapWireGuardObfuscationExpandButton()
>             .tapWireGuardObfuscationValueOnCell()
> ```

Nit: I'd prefer omitting `Value` here, as I accidentally read it as "tap the wireguard value on the cell".
___
*[`ios/MullvadVPNUITests/Networking/Networking.swift` line 67 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0R0c834WQXaeeL9Rr:-Nt0R0c834WQXaeeL9Rs:b-b4dty) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Networking/Networking.swift#L67)):*
> ```Swift
> 
>     /// Get configured ad serving domain as URL object
>     private static func getAdServingDomainURL() -> URL? {
> ```

Old code, but I think function should be based on `getAdServingDomain()` below, eg.:



_Code snippet:_
```Swift
guard let adServingDomain = getAdServingDomain(),
      let adServingDomainURL = URL(string: adServingDomain) else {
    XCTFail("Ad serving domain not configured")
    return nil
}
```
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 29 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0aLbU2jhJNOIzGp9b:-Nt0aLbV5CupLGrf4xb-:bvhkrl1) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L29)):*
> ```Swift
> 
>         let inAddressRow = app.otherElements[AccessibilityIdentifier.connectionPanelInAddressRow]
>         var capturedTexts: [String] = []
> ```

`capturedTexts` doesn't seem to be used for anything.
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 42 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0_qx01AHxw2LaCPs6:-Nt0_qx01AHxw2LaCPs7:b-6ghb40) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L42)):*
> ```Swift
>             if let currentText = inAddressRow.value as? String {
>                 // Skip initial label value with IP address only - no port or protocol
>                 if currentText.contains(" ") == false {
> ```

Nit: More common to use `guard` when returning early.
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 106 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0ZRUf19meNKTKx66D:-Nt0bMC-ES_dE6YI_pZp:bkvl4lj) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L106)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

Can we rename this to something more meaningful ?
</blockquote></details>

Perhaps the loops could be rewritten to include the index intead, eg.:



_Code snippet:_
```Swift
 for (attemptCount, attempt) in connectionAttempts {
    if attemptCount < 2 {
        XCTAssertEqual(attempt.protocolName, "UDP")
    } else if attemptCount == 2 {
        XCTAssertEqual(attempt.protocolName, "TCP")
    } else {
        XCTFail("Unexpected connection attempt")
    }
}
```
___
*[`ios/MullvadVPNUITests/Pages/VPNSettingsPage.swift` line 17 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0c3TQ13r3wHaLMduf:-Nt0c3TQ13r3wHaLMdug:b-53hq3t) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/VPNSettingsPage.swift#L17)):*
> ```Swift
>     }
> 
>     private func cellExpandButton(_ cellAccessiblityIdentifier: AccessibilityIdentifier) -> XCUIElement {
> ```

Nit: We look for `AccessibilityIdentifier.collapseButton`, but the function name is `cellEXPANDButton` 🙃

**Review by niklasberglund — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955)* status: 14 of 20 files reviewed, 15 unresolved discussions (waiting on @buggmagnet and @rablador)

___
*[`ios/MullvadVPN/Coordinators/ApplicationCoordinator.swift` line 770 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0LInl4fbtsg2xXwuV:-NtGwJUt7DiaLNTP5hCG:b-s6msf7) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPN/Coordinators/ApplicationCoordinator.swift#L770)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

How come this isn't set on the view controller inside `SettingsCoordinator` instead?
</blockquote></details>

I found it hard to follow the code here 😊 `SettingsCoordinator` sounds like a better place to set it but it's not a view. You mean it could be set inside `SettingsCoordinator` on `presentedViewController`?
___
*[`ios/MullvadVPN/View controllers/Settings/SettingsViewController.swift` line 58 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0MdVf8F2UUTccUWAI:-NtGyu6b0WCrhukHE446:bdq0ysl) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPN/View%20controllers/Settings/SettingsViewController.swift#L58)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

I know we discussed the issues with navigation bar buttons before, but do we really need to go through these hoops to make it work?
</blockquote></details>

This is to set accessibility identifier for the "Done" button in the upper right when in settings view 


`UIBarButtonItem` don't have any initialiser accepting accessibility identifier so need to store it as a variable and set `accessibilityIdentifier`
___
*[`ios/MullvadVPNUITests/RelayTests.swift` line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nsx5S36-9sQnC1BQLnd:-Nt0rsXkAzZUYpugTIy9:b-5bezxx) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/52dc88cd1c3a58735dcb7963254536e1e6d933a2/ios/MullvadVPNUITests/RelayTests.swift#L16)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

I'd say run `removeRules()` only when you needed, unless it's needed often enough to warrant a general inclusion. In general, I think we shouldn't proactively add stuff that we don't actually use.
</blockquote></details>

@buggmagnet true, but I think it makes sense that it's failing because of the workaround we use to get around the local network restriction? `8.8.8.8` is pointing to a local IP address. So it is treated as an external IP address even though the machine is on the local network.

@rablador sounds good to me! Have updated it to run only before and after the test that actually create firewall rules.
___
*[`ios/MullvadVPNUITests/RelayTests.swift` line 59 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0Q2EO4YPTrUMQyY8x:-NtH03X3AdYwdV10ayBI:b28m7bk) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/RelayTests.swift#L59)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

Nit: I'd prefer omitting `Value` here, as I accidentally read it as "tap the wireguard value on the cell".
</blockquote></details>

I like it. Have removed "Value" from the names.
___
*[`ios/MullvadVPNUITests/RelayTests.swift` line 80 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0I9o45D1ncdtgWx6f:-NtH07cV85xV_Wh5UdpE:b-xatl23) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/RelayTests.swift#L80)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

We can't really do this because the server IP might change in the future.
It's okay to leave it for now, but what we should probably be doing is either leveraging the JSON file that contains all the relays IPs  that we automatically download when we build the app, or do something similar to avoid having a static IP defined.
</blockquote></details>

See updated implementation where the test first connects to the relay to extract its IP address from the UI
___
*[`ios/MullvadVPNUITests/Networking/FirewallAPIClient.swift` line 83 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0J-tk2sNX1mw3jKML:-Nt0t3FnEN-RMbemr_vH:b-h7bvb) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75c053739a33b0314b13951a9dce04cfd1fb52da/ios/MullvadVPNUITests/Networking/FirewallAPIClient.swift#L83)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

Why did we remove that line ? What was not working ?
</blockquote></details>

It was added by mistake before. This request don't have any body so there's no need to specify a content type.
___
*[`ios/MullvadVPNUITests/Networking/Networking.swift` line 67 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0R0c834WQXaeeL9Rr:-NtH2yHhCh8WWL8Wq0Aw:b-cg8whm) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Networking/Networking.swift#L67)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

Old code, but I think function should be based on `getAdServingDomain()` below, eg.:
</blockquote></details>

Totally agree. I found that `getAdServingDomainURL` isn't even used though so removed it.
___
*[`ios/MullvadVPNUITests/Networking/Networking.swift` line 89 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0JhI14ge9AXFukO4N:-NtH0hfa9ajbEqmN2LWq:byxv7r2) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Networking/Networking.swift#L89)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

nit
The name doesn't read super well, I think we can rename this `alwaysReachableDomain` and get the same intended idea.

With this addition, the checks become more readable too IMHO
```swift
XCTAssertTrue(try canConnectSocket(host: alwaysReachableDomain(), port: "80"))
```

What do you think ?
</blockquote></details>

I like it 👍 See updated suggestion
___
*[`ios/MullvadVPNUITests/Networking/Networking.swift` line 154 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0KfoM2juaRflg2N-I:-NtH3AC0B-CFuNz5Qa7O:b-gayzjn) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Networking/Networking.swift#L154)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

nit 
`does not have`
</blockquote></details>

Thanks, I've always been confused about do and does 😄 Will probably make this mistake again
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 29 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0aLbU2jhJNOIzGp9b:-NtH3VNS1rPCGVKTqSk8:b-rkbpdg) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L29)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

`capturedTexts` doesn't seem to be used for anything.
</blockquote></details>

True, it was left from an earlier implementation and I forgot to remove it. Removed now.
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 31 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0LihU44p2w4FkHgBb:-NtH4Zbc5aMyTZpS_-Q4:b-qcittk) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L31)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

I understand what you're trying to do, but it looks really convoluted.
It's not performance critical code, so we should do `Date().timeIntervalSince(startTime)` instead for readability's sake
</blockquote></details>

Good idea 👍 Much easier to read
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 42 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0_qx01AHxw2LaCPs6:-NtH40IEAdEYpZDpyE2B:b-812wt5) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L42)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

Nit: More common to use `guard` when returning early.
</blockquote></details>

Updated to use `guard`
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 103 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0ZaMW-IkoURR7pRxq:-NtHHUXsDXkxW5bYnPa9:b-p8uvlg) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L103)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

nit
The logic in this function is a bit hard to follow, so the comments are very appreciated.
I'm wondering if we can make it easier to follow. I'll think about it.
</blockquote></details>

I totally agree and was thinking about this too. One way to make it easier to follow could be breaking out the asserts to a separate function accepting an array of protocol names and their order and the attempts array. Something like `assertAttemptOrder(connectionAttempts, ["UDP", "UDP", "TCP"]`. Then in this function it would just be something like:
```
            if connectionAttempts.count == 3 { // Expected retries flow
                assertAttemptOrder(connectionAttempts, ["UDP", "UDP", "TCP"]
                }
            } else if connectionAttempts.count == 2 { // Most of the times this incorrect flow is shown
                assertAttemptOrder(connectionAttempts, ["UDP", "TCP"]
                }
            } else {
                XCTFail("Unexpected number of connection attempts")
            }
```
___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 106 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0ZRUf19meNKTKx66D:-NtHJA5P9qTN6eTDhaRD:b-hqy30e) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L106)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

Perhaps the loops could be rewritten to include the index intead, eg.:
</blockquote></details>

I'm all for longer descriptive names 👍 Renamed `i` to `attemptIndex`. @rablador I like it. See updated syntax. Is this what you meant?
___
*[`ios/MullvadVPNUITests/Pages/VPNSettingsPage.swift` line 17 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0c3TQ13r3wHaLMduf:-NtHGFzT7FBuMbotzD3Q:b-hd5q7y) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPNUITests/Pages/VPNSettingsPage.swift#L17)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

Nit: We look for `AccessibilityIdentifier.collapseButton`, but the function name is `cellEXPANDButton` 🙃
</blockquote></details>

Yea this bothers me too. Actually the button isn't a collapse button OR expand button but both 😊 I can't come up with a better convention than naming them "expand collapse button". There should be a better name?

**Review by buggmagnet — APPROVED** [review]
:lgtm:

> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-:-NtQFQ_2CeIaKmwluntV:bnfp4nl)* status: all files reviewed, 12 unresolved discussions (waiting on @niklasberglund and @rablador)

___
*[`ios/MullvadVPNUITests/Pages/TunnelControlPage.swift` line 106 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-NtQF6LF7IoPgpudMCjb:-NtQF6LF7IoPgpudMCjc:b3xij97) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/0cb717062e40f45ed85f2eb48ca601b01623ae67/ios/MullvadVPNUITests/Pages/TunnelControlPage.swift#L106)):*
> ```Swift
>         // Should do three connection attempts but due to UI bug sometimes only two are displayed, sometimes all three
>         if connectionAttempts.count == 3 { // Expected retries flow
>             for (attemptIndex, attempt) in connectionAttempts.enumerated() {
> ```

That is much better now :thumbsup:

**Review by rablador — CHANGES_REQUESTED** [review]
:lgtm:

> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-:-NtUzU-C7VPBsGTl9z9a:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @buggmagnet and @niklasberglund)

___
*[`ios/MullvadVPN/Coordinators/ApplicationCoordinator.swift` line 770 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5955#-Nt0LInl4fbtsg2xXwuV:-NtUzy_e-UxiI4XfVSWA:b1keuvh) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fddf8d50c69c1432a067b950032b9cdb1ec31c78/ios/MullvadVPN/Coordinators/ApplicationCoordinator.swift#L770)):*
<details><summary><i>Previously, niklasberglund (Niklas Berglund) wrote…</i></summary><blockquote>

I found it hard to follow the code here 😊 `SettingsCoordinator` sounds like a better place to set it but it's not a view. You mean it could be set inside `SettingsCoordinator` on `presentedViewController`?
</blockquote></details>

Either that or `SettingsViewController`, which also resides there.
